core/src/main/java/com/linecorp/armeria/internal/server/QueryParamUtil.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.StringReader;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+
+import com.linecorp.armeria.common.AggregatedHttpRequest;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.QueryParams;
+import com.linecorp.armeria.common.QueryParamsBuilder;
+
+/**
+ * Parse query parameters.
+ */
+public abstract class QueryParamUtil {
+
+    /**
+     * Parse aggregated http request to query parameters.
+     */
+    public static QueryParams parse(AggregatedHttpRequest request) throws IOException {
```

Comment:
I think, we cannot use AnnotatedValueResolver.queryParamOf() because I want to QueryParamUtil will support:
- parse parameters from url
- parse parameters from body 
- parse upload files from multipart
AnnotatedValueResolver.queryParamOf() doesn't support all these features.

Reply:
Thanks for the explanation. 😄

I think `AnnotatedValueResolver.queryParamsOf()` can 
- parse parameters from URL.
- parse parameters from body.

But it cannot
- parse upload files from multipart

It's because we do not provide multipart support on Armeria at the moment. However, once we support it, annotated services and Servlet will use the same logic to parse that. Which means that we have to put the logic in the same place.

So what I recommended is, could you just copy the code into the `DefaultServletHttpRequest` and use it? I will send a follow up PR that removes the duplicate code. 😄